### PR TITLE
Ticket 1344 slicer

### PR DESCRIPTION
**Pull request by butlerpd**
This branch aims to allow the GUI to provide arbitrary box slicing (defined by xmax, xmin, ymax, ymin) and for boxes crossing Q=0,0 the option to "fold" or not.

NOTE: this is NOT ready for checking, but the first feature (allowing for + and - q to be plotted separately is needed by a user right now so providing this build as a stopgap till the fixes are finished.

NOTE2: This is written on the 4x series but the code (including the mistakes:-) seems to have been copied and pasted to the 5.0 so once we figure it out the changes should port easily to 5.x.

### CODE REVIEW

**Comment by butlerpd** [comment]
ready for testing on Win64

**Comment by butlerpd** [comment]
ready for testing on Win64

**Comment by butlerpd** [comment]
ready for testing on Win64

**Comment by butlerpd** [comment]
ready for testing on Win64

**Comment by butlerpd** [comment]
Looking at this again I think the rest of what would be nice to have will take much longer and this already fixes a bug and enhances the functionality so asking that it be reviewed and eventually merged.  Other enhancements can be added later.

File: src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py [hunk 1]
```diff
@@ -519,14 +534,36 @@ def send_to_fitting(self, fit=FIT_OPTIONS[0], file_list=None):
...
-        self.append_name.Enable(self.auto_save.IsChecked())
-        self.path.Enable(self.auto_save.IsChecked())
-        self.fitting_options.Enable(self.auto_save.IsChecked())
+        
+        """
+        _cb = evt.GetEventObject()
+        _cb_not_found = True
+        _param_index = next((i for i, v in enumerate(self.parameters)
+                                     if v[0] == "abs q"), None)
+        if (_param_index != None):
```
Comment (pkienzle) [inline src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py]: To compare with None use
```if _param_index is not None:```

File: src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py [hunk 2]
```diff
@@ -519,14 +534,36 @@ def send_to_fitting(self, fit=FIT_OPTIONS[0], file_list=None):
...
-    def on_auto_save_checked(self, evt=None):
+    def on_checbox_checked(self, evt=None):
         """
-        Enable/Disable auto append when checkbox is checked
+        All wx.checkbox events go to the same method so we have get the event
+        object to determine which box was checked and thus which code to
+        execute.  For this class the two checkbox events are used to:
+        
+        a) Enable/Disable auto append
+        b) togle between +/-q or abs q (|q|) slicer averaging/plotting
```
Comment (pkienzle) [inline src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py]: "togle" spelling.

File: src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py [hunk 3]
```diff
@@ -519,14 +534,36 @@ def send_to_fitting(self, fit=FIT_OPTIONS[0], file_list=None):
...
+        
+        """
+        _cb = evt.GetEventObject()
+        _cb_not_found = True
+        _param_index = next((i for i, v in enumerate(self.parameters)
+                                     if v[0] == "abs q"), None)
+        if (_param_index != None):
+            if _cb == self.abs_box:
+                _cb_not_found = False
+                (self.parameters[_param_index])[1] = _cb
```
Comment (pkienzle) [inline src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py]: It is odd that checking the checkbox for abs q will set the control associated in abs q to that checkbox.  Shouldn't the proper control have been put into the parameter list when the control was created?

File: src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py [hunk 4]
```diff
@@ -519,14 +534,36 @@ def send_to_fitting(self, fit=FIT_OPTIONS[0], file_list=None):
...
+                self.on_text_enter(wx.EVT_CHECKBOX)
+        if _cb == self.auto_save:
+            _cb_not_found = False
+            self.append_name.Enable(self.auto_save.IsChecked())
+            self.path.Enable(self.auto_save.IsChecked())
+            self.fitting_options.Enable(self.auto_save.IsChecked())
+        if _cb_not_found:
+            self.msg = "Checkbox generating event does not seem to be active."
+            logger.info(self.msg)
+
```
Comment (pkienzle) [inline src/sas/sasgui/guiframe/local_perspectives/plotting/parameters_panel_slicer.py]: Maybe use:
```python
cb = evt.GetEventObject()

# check for abs q folding
if cb == self.abs_box:
    for index, (key, control) in enumerate(self.parameters):
        if key == "abs q":
            self.parameters[index][1] = cb
            self.on_text_enter(wx.EVT_CHECKBOX)
            return
    # fall through to error logging if abs q not in parameter list

# check for auto save
if cb == self.auto_save:
    checked = self.auto_save.IsChecked()
    self.append_name.Enable(checked)
    self.path.Enable(checked)
    self.fitting_options.Enable(checked)
    return

# Fall through for other kinds of boxes
logger.info("Checkbox generating event does not seem to be active.")
```

File: src/sas/sasgui/guiframe/local_perspectives/plotting/boxSlicer.py [hunk 5]
```diff
@@ -26,19 +26,41 @@ def __init__(self, base, axes, color='black', zorder=3):
...
         self.x = 0.5 * min(math.fabs(self.base.data2D.xmax),
                            math.fabs(self.base.data2D.xmin))
-        self.y = 0.5 * min(math.fabs(self.base.data2D.xmax),
+        self.y = 0.5 * min(math.fabs(self.base.data2D.ymax),
+                           math.fabs(self.base.data2D.ymin))
+        self.xrange = self.base.data2D.xmax - self.base.data2D.xmin
+        self.yrange = self.base.data2D.ymax - self.base.data2D.ymin
+        if (self.base.data2D.xmax < 0) or (self.base.data2D.xmin > 0):
+            self.xmin = self.base.data2D.xmin + 0.25 * self.xrange
+            self.xmax = self.base.data2D.xmax + 0.25 * self.xrange
```
[the remark this answers is not in the document]
Comment (krzywon) [reply src/sas/sasgui/guiframe/local_perspectives/plotting/boxSlicer.py]: One other note: regardless of this change, the center of the box slicer is always set to Q = 0 with no way to change this in the GUI.

**Comment by butlerpd** [comment]
ready for testing on Win64

**Comment by wpotrzebowski** [comment]
Closing PR but keeping the branch

**Pull request closed without merging**